src/radical/pilot/agent/scheduler/continuous_summit.py

```diff
@@ -856,48 +864,44 @@ def _alloc_mpi(self, cud):
         if requested_lfs_per_process > lfs_per_node['size']:
             raise ValueError('Not enough LFS for the MPI-process')
 
-        if requested_cores > cores_per_node and \
-                cores_per_node % threads_per_proc and \
-                self._scattered is False:
-            raise ValueError('cannot allocate under given constrains')
-
-
 
         # set conditions to find the first matching node
-        is_first = True
-        is_last = False
+        is_first      = True
+        is_last       = False
         alloced_cores = 0
-        alloced_gpus = 0
-        alloced_lfs = 0
+        alloced_gpus  = 0
+        alloced_lfs   = 0
 
         slots = {'nodes': list(),
                  'cores_per_node': cores_per_node,
-                 'gpus_per_node': gpus_per_node,
-                 'lfs_per_node': lfs_per_node,
-                 'lm_info': self._lrms_lm_info,
-                 }
+                 'gpus_per_node' : gpus_per_node,
+                 'lfs_per_node'  : lfs_per_node,
+                 'lm_info'       : self._lrms_lm_info,
+                }
+
+        tag = cud.get('tag')
 
         # start the search
         for node in self.nodes:
 
-            node_uid = node['uid']
+            node_uid  = node['uid']
             node_name = node['name']
 
             # If unit has a tag, check if the tag is in the tag_history dict,
             # else it is a invalid tag, continue as if the unit does not have
             # a tag
             # If the unit has a valid tag, find the node that matches the
             # tag from tag_history dict
-            if tag and (tag in self._tag_history.keys()):
+            if tag and tag in self._tag_history:
```

Comment:
I'm not sure if checking if tag is not `None` and if it is in `tag_history` creates redundancy.
If tag is `None`, it would definitely not be in `tag_history`. Also, if tag has a value it can or cannot be in `tag_history`.

Isn't then:
```python
if tag in self._tag_history:
```
enough?